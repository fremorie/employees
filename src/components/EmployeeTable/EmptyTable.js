// @flow

import React from 'react';

type Props = {|
    nCols: number,
|};

const EmptyTable = ({nCols}: Props) => (
    <tr>
        <td colSpan={nCols}>
            Здесь пока ничего нет
        </td>
    </tr>
);

export default EmptyTable;
