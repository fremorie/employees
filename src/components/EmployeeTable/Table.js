// @flow

import React, {useState} from 'react';

import {Table} from 'react-bootstrap';

import {Edit} from '@material-ui/icons';

import type {Employee} from '../../types';

import EmptyTable from './EmptyTable';

const TableHeader = () => (
    <thead>
        <tr>
            <th></th>
            <th>Имя</th>
            <th>Фамилия</th>
            <th>Должность</th>
        </tr>
    </thead>
);

type EmployeeLineProps = {
    name: string,
    lastname: string,
    jobPosition: string,
    description?: string,
    orderNumber: number,
    onClick: () => void,
}

export const EmployeeLine = ({
    name,
    lastname,
    jobPosition,
    orderNumber,
    onClick,
}: EmployeeLineProps) => (
    <tr
        onClick={onClick}
        className="employeeLine"
    >
        <td width="50px">
            <span className="editIcon"><Edit fontSize="small" /></span>
            <span className="orderNumber">{orderNumber}</span>
        </td>
        <td>{name}</td>
        <td>{lastname}</td>
        <td>{jobPosition}</td>
    </tr>
);

type Props = {|
    employees: Employee[],
    openEmployeeInfo: (id: number) => void,
|};

export const TableBody = ({employees, openEmployeeInfo}: Props) => (
    <>
        {employees.map(({name, lastname, jobPosition, id}, i) =>
            (<EmployeeLine
                name={name}
                lastname={lastname}
                jobPosition={jobPosition}
                orderNumber={i + 1}
                onClick={() => openEmployeeInfo(id)}
                key={id}
            />)
        )}
    </>
);

const EmployeeTable = ({employees, openEmployeeInfo}: Props) => (
    <Table bordered striped hover variant="dark" responsive="sm">
        <TableHeader />
        <tbody>
            {employees.length
                ? <TableBody employees={employees} openEmployeeInfo={openEmployeeInfo} />
                : <EmptyTable nCols={4} />}
        </tbody>
    </Table>
);

export {EmployeeTable as Base};

export default EmployeeTable;
