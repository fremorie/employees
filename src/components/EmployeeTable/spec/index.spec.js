import React from 'react';
import {mount} from 'enzyme';

import {TableBody, EmployeeLine} from '../Table';
import EmptyTable from '../EmptyTable';

describe('EmployeeTable', () => {
    describe('EmptyTable', () => {
        it('рендерится', () => {
            const component = mount(
                <EmptyTable nCols={3} />
            );

            expect(component).toMatchSnapshot();
        });
    });

    describe('TableBody', () => {
        it('рендерится', () => {
            const props = {
                employees: [
                    {name: 'name', lastname: 'lastname', jobPosition: 'jobPosition', description: 'description', id: 1},
                    {name: 'name1', lastname: 'lastname1', jobPosition: 'jobPosition1', description: 'description1', id: 2},
                ],
                openEmployeeInfo: jest.fn(),
            };

            const component = mount(
                <TableBody {...props} />
            );

            expect(component.find('.employeeLine')).toHaveLength(2);
        });
    });

    describe('EmployeeLine', () => {
        it('рендерится', () => {
            const props = {
                name: 'name',
                lastname: 'lastname',
                jobPosition: 'jobPosition',
                orderNumber: 'orderNumber',
                onClick: jest.fn(),
            };

            const component = mount(
                <EmployeeLine {...props} />
            );

            expect(component.find('td')).toHaveLength(4);
            expect(component.find('tr')).toHaveLength(1);
        });
    });
});
