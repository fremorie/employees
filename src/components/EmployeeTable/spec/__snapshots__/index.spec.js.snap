// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeTable EmptyTable рендерится 1`] = `
ReactWrapper {
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "getWrappingComponentRenderer": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
    "wrapInvoke": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "nCols": 3,
    },
    "ref": null,
    "rendered": Object {
      "instance": <tr>
        <td
          colspan="3"
        >
          Здесь пока ничего нет
        </td>
      </tr>,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <td
          colSpan={3}
        >
          Здесь пока ничего нет
        </td>,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": <td
            colspan="3"
          >
            Здесь пока ничего нет
          </td>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Здесь пока ничего нет",
            "colSpan": 3,
          },
          "ref": null,
          "rendered": Array [
            "Здесь пока ничего нет",
          ],
          "type": "td",
        },
      ],
      "type": "tr",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "nCols": 3,
      },
      "ref": null,
      "rendered": Object {
        "instance": <tr>
          <td
            colspan="3"
          >
            Здесь пока ничего нет
          </td>
        </tr>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <td
            colSpan={3}
          >
            Здесь пока ничего нет
          </td>,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": <td
              colspan="3"
            >
              Здесь пока ничего нет
            </td>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Здесь пока ничего нет",
              "colSpan": 3,
            },
            "ref": null,
            "rendered": Array [
              "Здесь пока ничего нет",
            ],
            "type": "td",
          },
        ],
        "type": "tr",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__linkedRoots__): Array [],
  Symbol(enzyme.__unrendered__): <EmptyTable
    nCols={3}
  />,
  Symbol(enzyme.__updatedBy__): null,
}
`;
